src/commonMain/kotlin/org/kosat/heuristics/Restarter.kt

```diff
@@ -0,0 +1,163 @@
+            restart()
+        }
+    }
+
+    // for each literal provides a list of clauses containing it (for 'x' it's in pos x, for 'not x' in pos solver.varsNumber + x)
+    private var litOccurrence = mutableListOf<MutableList<Int>>()
+    // vars.mapIndexed { ind, _ -> clauses.mapIndexed { ind, _ -> ind}.filter { clauses[it].contains(ind) || clauses[it].contains(-ind) }.toMutableList()}
+
+    // return position of literal in occurrence array
+    private fun litPos(lit: Int): Int {
+        return if (lit >= 0) {
+            lit * 2 - 1 // odd indexes
+        } else {
+            -lit * 2 // even indexes
+        }
+    }
+
+    fun countOccurrence() {
+        litOccurrence.clear()
+        for (ind in 1..(2 * solver.varsNumber + 1)) {
+            litOccurrence.add(mutableListOf())
+        }
+        solver.clauses.forEachIndexed { ind, clause ->
+            clause.forEach { lit -> litOccurrence[litPos(lit)].add(ind) }
+        }
+    }
+
+    // remove subsumed clauses
+    private fun removeSubsumedClauses() {
+        val uselessClauses = mutableSetOf<Int>()
+        val duplicateClauses = mutableSetOf<Int>()
+        val markedClauses = MutableList(solver.clauses.size) { false }
+
+        // going from the end because smaller clauses appear after big one
+        for (ind in solver.clauses.lastIndex downTo 0) {
+            if (!markedClauses[ind]) {
+                findSubsumed(ind).forEach {
+                    if (!markedClauses[it]) {
+                        if (ind.clauseSize() < it.clauseSize()) {
+                            markedClauses[it] = true
+                            uselessClauses.add(it)
+                        } else if (ind.clauseSize() == it.clauseSize()) {
+                            duplicateClauses.add(it)
+                        }
+                    }
+                }
+            }
+        }
+
+        val copiedClauses = duplicateClauses.map { solver.clauses[it] }
+
+        solver.clauses.removeAll(uselessClauses.map { solver.clauses[it] })
+        // remove duplicate clauses and leave 1 copy of each
+        solver.clauses.removeAll(copiedClauses)
+        solver.clauses.addAll(copiedClauses)
+        countOccurrence()
+        updateSig()
+    }
+
+    private val hash = MutableList(2 * solver.varsNumber + 1) { 1L.shl(it % 64) }
+
+    private fun countSig(clause: Int): Long {
+        var sz = 0L
+        solver.clauses[clause].forEach { lit -> sz = sz.or(hash[litPos(lit)]) }
+        return sz
+    }
+
+    private fun Int.clauseSize() = solver.clauses[this].size
+
+    private var clauseSig = mutableListOf<Long>()
+
+    fun updateSig() {
+        clauseSig = solver.clauses.mapIndexed { ind, _ -> countSig(ind) }.toMutableList()
+    }
+
+    override fun addVariable() {
+        for (i in 0..1) {
+            hash.add(1L.shl(hash.size % 64))
+            litOccurrence.add(mutableListOf())
+        }
+    }
+
+    private fun findSubsumed(clause: Int): Set<Int> {
```

Comment:
Add docs on what does this function return. Set of what? Clause ids?—then why not just `List<Clause>`? Literals?—then why `Set`?